### PR TITLE
docs: add ast-spec, type-utils docs with docusaurus-plugin-typedoc

### LINKED ISSUE #2726: Documentation for AST node types
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

not applicable

**Expected Result**

not applicable

**Actual Result**

not applicable

**Additional Info**

This is a meta issue about documentation.

I couldn't find any information about the AST node types, especially the ones that are Typescript-specific. For example, if I'm extending a rule and I come across a place where it's dealing with `TSTypeParameterInstantiation` - how do I know what that means? Ideally, it would be great to have a gallery with code snippets (examples) for every available node type. I know that's a great effort, but even starting to add some JSDoc comments with very brief explanations or short snippets to the code would help a lot.

Here's an example of what I'm thinking of:

**Before**
```ts
export interface CallExpression extends CallExpressionBase {
  type: AST_NODE_TYPES.CallExpression;
}
```

**After**
```ts
/**
 * A function/method call.
 * @example function1<Type1>(arg1);
 */
export interface CallExpression extends CallExpressionBase {
  type: AST_NODE_TYPES.CallExpression;
}
```

Of course, that's an easy example, everyone knows what a call expression is. But for more complex types, this would make writing rules significantly easier.

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.6.0` |
| `TypeScript`                           | `3.4.3` |
| `node`                                 | `12.19.0` |

### LINKED ISSUE #8093: Docs: Document @typescript-eslint/type-utils
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

https://www.npmjs.com/package/@typescript-eslint/type-utils says:

> 👉 See https://typescript-eslint.io/packages/type-utils for documentation on this package.

https://typescript-eslint.io/packages/type-utils is a 404.

Additionally, #8089 raises a good point: it's not clear whether `@typescript-eslint/type-utils` is part of our public API...
* If it is: we should document it and fill in the 404 page
* If it isn't: we should change its README.md to mention it

### Affected URL(s)

https://typescript-eslint.io/packages/type-utils and/or https://www.npmjs.com/package/@typescript-eslint/type-utils

### LINKED ISSUE #9298: Docs: Add generated documentation on packages exports to the site
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

Splitting out of #8093: in addition to having a landing page for the `utils` and `type-utils` packages on the website, I think it'd be nice to _additionally_ have generated docs on their actual exports. It's more discoverable / searchable than having to crawl through TypeScript definitions.

From https://github.com/typescript-eslint/typescript-eslint/issues/8093#issuecomment-2152422374, I have a PoC over at https://deploy-preview-9293--typescript-eslint.netlify.app/packages/type-utils. What do you all think about having this for, say, the packages with standalone exports folks would typically use:
* `type-utils`
* `utils`

### Affected URL(s)

https://typescript-eslint.io/packages/type-utils (once it exists)

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #2726; fixes #9298
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Sets up https://typedoc-plugin-markdown.org/plugins/docusaurus to auto-generate TypeDoc docs from `packages/type-utils/src/index.ts` into `docs/package/type-utils/api`.

💖

### CODE REVIEW

File: packages/ast-spec/src/ast-node-types.ts [hunk 1]
```diff
@@ -88,9 +88,9 @@ export enum AST_NODE_TYPES {
   WhileStatement = 'WhileStatement',
   WithStatement = 'WithStatement',
   YieldExpression = 'YieldExpression',
-  /**
-   * TS-prefixed nodes
-   */
+
+  // TS_prefixed nodes
```
Comment (JoshuaKGoldberg) [inline packages/ast-spec/src/ast-node-types.ts:92]: The `/** ... */` JSDoc style comment was getting picked up as a comment on specifically `TSAbstractAccessorProperty`.

File: packages/ast-spec/src/declaration/VariableDeclaration/spec.ts [hunk 2]
```diff
@@ -11,7 +11,8 @@ export interface LetOrConstOrVarDeclaration extends BaseNode {
   /**
    * The variables declared by this declaration.
    * Note that there may be 0 declarations (i.e. `const;`).
-   * ```
+   * @example
+   * ```ts
```
Comment (JoshuaKGoldberg) [inline packages/ast-spec/src/declaration/VariableDeclaration/spec.ts:15]: I tried to make sure all <code>```</code>s had a language and were in a proper `@example`, so they'd play nicely with docs generation. But multiline ones expose a bug in the markdown table output 🫤:

Screenshot of the LetOrConstVariableDeclaration where only the first line of the example code blocks are actual markdown codeblocks

Comment (JoshuaKGoldberg) [reply packages/ast-spec/src/declaration/VariableDeclaration/spec.ts:15]: Resolved by splitting into individual `@example`s. 🤷

Comment (JoshuaKGoldberg) [reply packages/ast-spec/src/declaration/VariableDeclaration/spec.ts:15]: Some of them are still a little messy... but I don't know the right way to get around this. 😬 putting in review.

**Review by bradzacher — COMMENTED** [review]
https://deploy-preview-9293--typescript-eslint.netlify.app/packages/utils/generated/namespaces/ASTUtils/

Bug: This page loses the sidebar (and is not listed in the sidebar at all).

---

screenshot of TOC

Question: can we change the TOC so that we show one fewer level of headings on these pages? It's pretty low value showing the `Parameters`, `Returns` etc headings in the TOC -- i'd say it's even just noisy / confusing.

**Comment by JoshuaKGoldberg** [comment]
Yeah @bradzacher that makes sense. I was excited about getting the `utils` package in there but it's a bit more complicated and I think we'll need to look more critically at its docs.

The presence of multiple / nested namespaces makes the doc gen trickier. I tried much of https://typedoc-plugin-markdown.org/docs/options/file-options but couldn't figure out how to collapse it all into one page.

For now, I just removed `utils` from the array of generated package docs.

File: packages/ast-spec/src/element/TSEnumMember/spec.ts [hunk 3]
```diff
@@ -19,12 +19,18 @@ interface TSEnumMemberBase extends BaseNode {
  * this should only really happen in semantically invalid code (errors 1164 and 2452)
  *
  * VALID:
+ * @example
```
Comment (Josh-Cena) [inline packages/ast-spec/src/element/TSEnumMember/spec.ts:22]: This doesn't work. A JSDoc comment cannot have duplicate keys and here there are two `@example` keys. I suggest just using plain descriptions instead.

Comment (JoshuaKGoldberg) [reply packages/ast-spec/src/element/TSEnumMember/spec.ts:22]: Huh, TIL - thanks! I just reduced it down to the one `@example` to keep the change minimal.

File: packages/ast-spec/src/declaration/VariableDeclaration/spec.ts [hunk 4]
```diff
@@ -20,16 +21,25 @@ export interface LetOrConstOrVarDeclaration extends BaseNode {
...
    * declare const x = 1;
    * ```
    */
   declare: boolean;
   /**
    * The keyword used to declare the variable(s)
    * ```
+   * @example
```
Comment (Josh-Cena) [inline packages/ast-spec/src/declaration/VariableDeclaration/spec.ts:33]: Same here: not sure why we want to split them up (and I think there's a stray `` ``` `` above)

**Comment by JoshuaKGoldberg** [comment]
Merging after talking privately & confirming that:

* The docs here are so big that it's super hard to actually review
* This is generally good to go otherwise

**Pull request merged**